### PR TITLE
Introduce pure BF16 training with Kaha summation - (torch-optimi package)

### PR DESCRIPTION
**Pull request by vince62s**
This PR also removes deprecated apex fusedadam which is no longer needed.

Pure BF16 is the fastest way and lowest Memory footprint way to train with the same accuracy as fp16/bf16 with AMP (fp32 weights)

### CODE REVIEW

File: eole/utils/optimizers.py [hunk 1]
```diff
@@ -41,17 +41,22 @@ def build_torch_optimizer(model, config):
     """
     params = [p for p in model.parameters() if p.requires_grad]
     betas = [config.adam_beta1, config.adam_beta2]
+    print(config.use_amp, optimi_available)
```
Comment (francoishernandez) [inline eole/utils/optimizers.py:44]: To remove?

File: eole/utils/optimizers.py [hunk 2]
```diff
@@ -41,17 +41,22 @@ def build_torch_optimizer(model, config):
...
+    if config.use_amp or not optimi_available:
+        optim = torch_optim
+    else:
+        optim = torch_optimi
```
Comment (francoishernandez) [inline eole/utils/optimizers.py:48]: `torch_optim` / `torch_optimi` difference not super readable, maybe just use directly `torch.optim` and `optimi`?
The fact that `torch.optim` is also used directly does not help either.

File: eole/utils/optimizers.py [hunk 3]
```diff
@@ -41,17 +41,22 @@ def build_torch_optimizer(model, config):
...
-        optimizer = optim.Adagrad(
+        optimizer = torch.optim.Adagrad(
             params,
             lr=config.learning_rate,
             initial_accumulator_value=config.adagrad_accumulator_init,
             weight_decay=config.weight_decay,
         )
     elif config.optim == "adadelta":
-        optimizer = optim.Adadelta(params, lr=config.learning_rate, weight_decay=config.weight_decay)
+        optimizer = torch.optim.Adadelta(params, lr=config.learning_rate, weight_decay=config.weight_decay)
```
Comment (francoishernandez) [inline eole/utils/optimizers.py:59]: Should we check against use_amp /optimi for such cases, to make it clear that such optimizers won't use the `optimi` based codepath?

File: eole/utils/optimizers.py [hunk 4]
```diff
@@ -88,46 +93,10 @@ def build_torch_optimizer(model, config):
                 dense.append(param)
         optimizer = MultipleOptimizer(
             [
-                optim.Adam(dense, lr=config.learning_rate, betas=betas, eps=1e-8),
-                optim.SparseAdam(sparse, lr=config.learning_rate, betas=betas, eps=1e-8),
+                optim.Adam(dense, lr=config.learning_rate, betas=betas, eps=config.adam_eps),
+                torch.optim.SparseAdam(sparse, lr=config.learning_rate, betas=betas, eps=config.adam_eps),
```
Comment (francoishernandez) [inline eole/utils/optimizers.py:97]: Same as above, we might have weird stuff if `optim` == `optimi` here.

**Review by francoishernandez — COMMENTED** [review]
Nice one. This will also help simplify the docker stack, as building custom apex won't be necessary anymore!
